### DISCUSSION TITLE
Help deserializing enum from class

### DISCUSSION THREAD

**Discussion by mfallavol** [nlohmann/json#2620] [Q&A]
I'm sorry to ask this but I just can't seem to figure out how to de-serialize an enum that is part of a class/struct.  Here is an example of what I'm trying to do:

```cpp
#include <iostream>
#include "nlohmann/json.hpp"

using json = nlohmann::json;

struct MyEnum
{
	enum Enum
	{
		Zero,
		One,
		Two,
		Three,
	} e;

	MyEnum	operator=(Enum o)		{ e = o; return *this; }
	MyEnum	operator=(MyEnum o)		{ e = o.e; return *this; }
	bool	operator==(Enum o)		{ return e == o; }
	bool	operator==(MyEnum o)	{ return e == o.e; }
			operator Enum() const	{ return e; }

	NLOHMANN_JSON_SERIALIZE_ENUM( Enum, {
		{ Zero,		"Zero"},
		{ One,		"One"},
		{ Two,		"Two"},
		{ Three,	"Three"},
	})
};

int
main(int argc, char **argv)
{
	json	inData;
	MyEnum	me;
	
	inData["number"] = "Two";
		   
	me = inData["number"].get<MyEnum::Enum>();
	std::cout << "me " << me << std::endl;
}
```

The above compiles with g++-9 on ubuntu 18.04 using: g++ -o enumTest  -std=c++2a enumTest.cpp
The result is an exception:
terminate called after throwing an instance of 'nlohmann::detail::type_error'
  what():  [json.exception.type_error.302] type must be number, but is string
Aborted (core dumped)

I understand this has something to do with the namespaces and falls under this statement in the documentation:

> NLOHMANN_JSON_SERIALIZE_ENUM() MUST be declared in your enum type's namespace (which can be the global namespace), or the library will not be able to locate it and it will default to integer serialization.

but I just don't know how to fix it.
I'd appreciate it if someone more adept at C++ would take a look.

**Comment by maddanio** [comment]
I didn't know NLOHMANN_JSON_SERIALIZE_ENUM, I will have a look at it. In the meantime here is how we serialize any enums using [magic_enum](https://github.com/Neargye/magic_enum):
```
#include <magic_enum.hpp>
#include <nlohmann/json.hpp>
#include <boost/algorithm/string.hpp>
#include <boost/format.hpp>

#include <stdexcept>

    template <typename T>
    std::enable_if_t<std::is_enum<T>::value, void>
    from_json(const nlohmann::json &j, T & obj)
    {
        auto enum_str = j.get<std::string>();
        for (auto&& entry : magic_enum::enum_entries<T>())
            if (boost::iequals(entry.second, enum_str))
            {
                obj = entry.first;
                return;
            }
        throw std::runtime_error(str(boost::format("Unkown enum: %s")%enum_str));
    }

    template <typename T>
    std::enable_if_t<std::is_enum<T>::value, void>
    to_json(nlohmann::json &j, const T & obj)
    {
        j = std::string{magic_enum::enum_name(obj)};
    }
```

**Comment by maddanio** [comment]
Ok, in your example I am not sure why the Enum class exists. You can just declare that enum globally and then it would work. if you want the class you will have to define serialization for it as well as the macro only does this for the actual enum.

**Comment by nlohmann** [comment]
Do you need further assistance with this issue?
